File: tests/fixtures/rust/type_inference/generics/generics.rs
use std::fmt::Display;

pub fn print_pair<T: Display, U: Display>(a: T, b: U) -> String {
    format!("{}: {}", a, b)
}

pub struct Container<T> {
    pub value: T,
}

impl<T: Clone> Container<T> {
    pub fn duplicate(&self) -> Container<T> {
        Container {
            value: self.value.clone(),
        }
    }
}

pub fn nested_generic() -> Vec<Option<i32>> {
    vec![Some(1), None, Some(3)]
}

pub fn identity<T>(x: T) -> T {
    x
}

pub fn wrap_in_vec<T>(item: T) -> Vec<T> {
    vec![item]
}

pub struct Pair<A, B> {
    pub first: A,
    pub second: B,
}

impl<A, B> Pair<A, B> {
    pub fn new(first: A, second: B) -> Self {
        Pair { first, second }
    }

    pub fn swap(self) -> Pair<B, A> {
        Pair {
            first: self.second,
            second: self.first,
        }
    }
}

pub fn collect_to_map<K, V, I>(iter: I) -> std::collections::HashMap<K, V>
where
    K: std::hash::Hash + Eq,
    I: IntoIterator<Item = (K, V)>,
{
    iter.into_iter().collect()
}

pub fn main() {
    let pair = print_pair(42, "hello");
    let container = Container { value: String::from("test") };
    let dup = container.duplicate();
    let nested = nested_generic();
    let wrapped = wrap_in_vec(10);
    let p = Pair::new(1, "one");
    let swapped = p.swap();
}

File: tests/fixtures/kotlin/type_inference/generics/generics.kt
fun <T> identity(x: T): T = x

fun <T> wrapInList(item: T): List<T> = listOf(item)

fun <K, V> toMap(key: K, value: V): Map<K, V> = mapOf(key to value)

class Container<T>(val value: T) {
    fun duplicate(): Container<T> = Container(value)
}

class Pair<A, B>(val first: A, val second: B) {
    fun swap(): Pair<B, A> = Pair(second, first)
}

fun <T : Comparable<T>> maxOf(a: T, b: T): T = if (a >= b) a else b

fun main() {
    val id: String = identity("test")
    val list = wrapInList("item")
    val map = toMap("key", 100)
    val container = Container("value")
    val dup = container.duplicate()
    val p = Pair(1, "one")
    val swapped = p.swap()
    val biggest: Int = maxOf(10, 20)
    println("$id $list $map $dup $swapped $biggest")
}
